src/wp-includes/option.php

```diff
@@ -2144,20 +2144,50 @@ function update_network_option( $network_id, $option, $value ) {
 	 */
 	$value = apply_filters( "pre_update_site_option_{$option}", $value, $old_value, $option, $network_id );
 
+	/*
+	 * To get the actual raw old value from the database, any existing pre filters need to be temporarily disabled.
+	 * Immediately after getting the raw value, they are reinstated.
+	 * The raw value is only used to determine whether a value is present in the database. It is not used anywhere
+	 * else, and is not passed to any of the hooks either.
+	 */
+	global $wp_filter;
+
+	$default_value_hook = is_multisite() ? "default_site_option_{$option}" : "default_option_{$option}";
+	$default_value      = apply_filters( $default_value_hook, false, $option, $network_id );
+	$pre_option_hook    = is_multisite() ? "pre_site_option_{$option}" : "pre_option_{$option}";
+
+	if ( has_filter( $pre_option_hook ) ) {
+		$old_filters = $wp_filter[ $pre_option_hook ];
+		unset( $wp_filter[ $pre_option_hook ] );
+
+		$raw_old_value = is_multisite() ? get_network_option( $network_id, $option ) : get_option( $option, $default_value );
+
+		$wp_filter[ $pre_option_hook ] = $old_filters;
+	} else {
+		$raw_old_value = $old_value;
+	}
```

Comment:
Sharing from my Slack message:
I think the new logic is still not correct, and is functionally different from what I proposed in https://github.com/WordPress/wordpress-develop/pull/5434#pullrequestreview-1665030594. We cannot only check for `pre_option_{$option}` when `! is_multisite()`. The function calls are technically nested, `get_network_option()` calls `get_option()` so both `pre_site_option_{$option}` and `pre_option_{$option}` fire. The current code still doesn't address this.

If for example an option has a `pre_site_option_{$option}` filter but the function is called on a non-multisite, the current code wouldn't behave correctly.

Reply:
I think something similar applies to the default value hook. We shouldn't skip the `default_site_option_{$option}` hook, otherwise the `$raw_old_value` retrieved may be incorrect.

Currently, on non-multisite, the logic will run the `default_option_{$option}` filter and then pass the result to `get_option()`, which makes no sense, since that function will run that filter anyway.

Reply:
I agree @felixarntz. [9a2dccd](https://github.com/WordPress/wordpress-develop/pull/5434/commits/9a2dccdae048c028d6503bf5fd8a4c7b9264e926), aims to address both the `pre_` and `default_` filter hooks.